overlay/vault-push-roles.nix

```diff
@@ -0,0 +1,197 @@
+# Generate and push approles to vault
+{ writeShellScriptBin, jq, vault, coreutils, lib }:
+# Inputs: a flake with `nixosConfigurations`
+
+# Usage:
+# apps.x86_64-linux.vault-push-approles = { type = "app"; program = "${pkgs.vault-push-approles self}/bin/vault-push-approles"; }
+
+{ nixosConfigurations ? { }, ... }: rec {
+  # Overrideable functions
+  # Usage examples:
+  # pkgs.vault-push-approles self { approleCapabilities.aquarius-albali-borgbackup = [ "read" "write" ]; }
+  /* pkgs.vault-push-approles self (final: prev: {
+       approleCapabilitiesFor = { approleName, namespace, ... }@params:
+         if namespace == "albali" then [ "read" "write" ] else prev.approleCapabilitiesFor params;
+     })
+  */
+  # pkgs.vault-push-approles { } { extraApproles = [ { ... } ] }
+
+  overrideable = final: {
+    extraApproles = [ ];
+
+    renderJSON = name: content:
+      builtins.toFile "${name}.json" (builtins.toJSON content);
+
+    approleParams = {
+      secret_id_ttl = "";
+      token_num_uses = 0;
+      token_ttl = "20m";
+      token_max_ttl = "30m";
+      secret_id_num_uses = 0;
+    };
+
+    mkApprole = { approleName, name, ... }:
+      final.renderJSON "approle-${approleName}"
+      (final.approleParams // { token_policies = [ name ]; });
+
+    approleCapabilities = { };
+
+    approleCapabilitiesFor = { approleName, ... }:
+      final.approleCapabilities.${approleName} or [ "read" ];
+
+    mkPolicy = { approleName, name, vaultPathPrefix, namespace, ... }@params: let
+      splitPrefix = builtins.filter builtins.isString (builtins.split "/" vaultPathPrefix);
+
+      insertAt = lst: index: value: (lib.lists.take index lst) ++ [value] ++ (lib.lists.drop index lst);
+
+      makePrefix = value: builtins.concatStringsSep "/" (insertAt splitPrefix 1 value);
+
+      metadataPrefix = makePrefix "metadata";
+      dataPrefix = makePrefix "+";
+    in
+      final.renderJSON "policy-${approleName}" {
+        path = [
+          {
+            "${metadataPrefix}/${namespace}/${name}/*" =
+              [{ capabilities = [ "list" ]; }];
+          }
+          {
+            "${dataPrefix}/${namespace}/${name}/*" =
+              [{ capabilities = final.approleCapabilitiesFor params; }];
+          }
+        ];
+      };
```

Comment:
I'm not sure when the policies need to include `buildkite/*` stuff (like `buildkite-agent-public` does).

Reply:
For now, it can be mitigated with

```
      vault-push-roles = pkgs.vault-push-roles self (final: prev: {
        mkPolicy = { approleName, ... }@params:
          if !isNull (builtins.match ".*buildkite.*" approleName) then {
            path = (prev.mkPolicy params).path
                ++ (prev.mkPolicy (params // {
                  vaultPathPrefix = "buildkite";
                  namespace =
                    if !isNull (builtins.match ".*private.*" approleName) then
                      "serokell-private"
                    else
                      "serokell";
                })).path;
          } else
            prev.mkPolicy params;
      });
```

in aquarius-infra.